File: internal/dtls13/schedule.go
package dtls13

import (
	"crypto/hkdf"
	"crypto/hmac"
	"hash"
)

type keySchedule struct {
	suite           cipherSuite
	transcript      hash.Hash
	master          []byte
	clientHandshake []byte
	serverHandshake []byte
}

func newKeySchedule(id uint16, shared, helloTranscript []byte) (*keySchedule, error) {
	suite, err := suiteFor(id)
	if err != nil {
		return nil, err
	}
	if len(shared) == 0 {
		return nil, errKeyMaterial
	}
	zeros := make([]byte, suite.hash().Size())
	early, err := hkdf.Extract(suite.hash, zeros, zeros)
	if err != nil {
		return nil, err
	}
	derived, err := expandLabel(suite.hash, early, "derived", suite.hash().Sum(nil), len(zeros))
	if err != nil {
		return nil, err
	}
	handshake, err := hkdf.Extract(suite.hash, shared, derived)
	if err != nil {
		return nil, err
	}
	s := &keySchedule{suite: suite, transcript: suite.hash()}
	_, _ = s.transcript.Write(helloTranscript)
	s.clientHandshake, err = s.derive(handshake, "c hs traffic")
	if err != nil {
		return nil, err
	}
	s.serverHandshake, err = s.derive(handshake, "s hs traffic")
	if err != nil {
		return nil, err
	}
	derived, err = expandLabel(suite.hash, handshake, "derived", suite.hash().Sum(nil), len(zeros))
	if err != nil {
		return nil, err
	}
	s.master, err = hkdf.Extract(suite.hash, zeros, derived)
	if err != nil {
		return nil, err
	}
	return s, nil
}

func (s *keySchedule) derive(secret []byte, label string) ([]byte, error) {
	return expandLabel(s.suite.hash, secret, label, s.transcript.Sum(nil), s.transcript.Size())
}

func (s *keySchedule) write(m handshakeMessage) error {
	wire, err := m.transcript()
	if err != nil {
		return err
	}
	_, _ = s.transcript.Write(wire)
	return nil
}

func (s *keySchedule) finished(secret []byte) ([]byte, error) {
	if len(secret) != s.transcript.Size() {
		return nil, errKeyMaterial
	}
	key, err := expandLabel(s.suite.hash, secret, "finished", nil, s.transcript.Size())
	if err != nil {
		return nil, err
	}
	h := hmac.New(s.suite.hash, key)
	_, _ = h.Write(s.transcript.Sum(nil))
	return h.Sum(nil), nil
}

func (s *keySchedule) applicationSecrets() (client, server []byte, err error) {
	client, err = s.derive(s.master, "c ap traffic")
	if err != nil {
		return nil, nil, err
	}
	server, err = s.derive(s.master, "s ap traffic")
	return client, server, err
}

func nextTrafficSecret(id uint16, secret []byte) ([]byte, error) {
	suite, err := suiteFor(id)
	if err != nil {
		return nil, err
	}
	if len(secret) != suite.hash().Size() {
		return nil, errKeyMaterial
	}
	return expandLabel(suite.hash, secret, "traffic upd", nil, suite.hash().Size())
}

func retryTranscript(id uint16, firstClientHello, retry, secondClientHello []byte) ([]byte, error) {
	suite, err := suiteFor(id)
	if err != nil {
		return nil, err
	}
	h := suite.hash()
	_, _ = h.Write(firstClientHello)
	w := wireWriter{}
	w.uint8(msgMessageHash)
	w.vector24(h.Sum(nil))
	w.data = append(w.data, retry...)
	w.data = append(w.data, secondClientHello...)
	return w.result()
}

File: internal/dtls13/testdata/schedule.py
"""Regenerate schedule.json using Python's independent HMAC implementation.

The synthetic transcripts exercise HKDF and transcript framing, not message
validation. Real handshakes are checked separately against independent peers.
"""

import hashlib
import hmac
import json
from pathlib import Path


def vector(name):
    digest = getattr(hashlib, name)
    size = digest().digest_size

    def mac(key, value):
        return hmac.digest(key, value, name)

    def expand(secret, label, context):
        label = b"dtls13" + label
        info = size.to_bytes(2, "big") + bytes([len(label)]) + label
        info += bytes([len(context)]) + context
        return mac(secret, info + b"\x01")

    def message(kind, body):
        return bytes([kind]) + len(body).to_bytes(3, "big") + body

    shared = bytes(range(32))
    first_hello = message(1, b"first client hello")
    retry = message(2, b"retry cookie")
    second_hello = message(1, b"second client hello")
    hello = message(1, b"client hello") + message(2, b"server hello")
    empty_hash = digest().digest()
    early = mac(bytes(size), bytes(size))
    handshake = mac(expand(early, b"derived", empty_hash), shared)
    master = mac(expand(handshake, b"derived", empty_hash), bytes(size))
    client_handshake = expand(handshake, b"c hs traffic", digest(hello).digest())
    server_handshake = expand(handshake, b"s hs traffic", digest(hello).digest())
    transcript = hello + message(8, b"extensions") + message(11, b"certificate")
    transcript += message(15, b"certificate verify")
    server_finished = mac(expand(server_handshake, b"finished", b""), digest(transcript).digest())
    transcript += message(20, server_finished)
    client_finished = mac(expand(client_handshake, b"finished", b""), digest(transcript).digest())
    client_application = expand(master, b"c ap traffic", digest(transcript).digest())
    server_application = expand(master, b"s ap traffic", digest(transcript).digest())
    result = {
        "shared": shared, "hello": hello, "master": master,
        "client_handshake": client_handshake, "server_handshake": server_handshake,
        "server_finished": server_finished, "client_finished": client_finished,
        "client_application": client_application, "server_application": server_application,
        "client_update": expand(client_application, b"traffic upd", b""),
        "retry_transcript": message(254, digest(first_hello).digest()) + retry + second_hello,
    }
    return {key: value.hex() for key, value in result.items()}


if __name__ == "__main__":
    Path(__file__).with_suffix(".json").write_text(
        json.dumps({name: vector(name) for name in ("sha256", "sha384")}, indent=2) + "\n",
        encoding="utf-8",
    )
